fix syskit support

doudou opened the pull request:
These are commits that were exclusively in https://github.com/rock-gazebo/drivers-transformer. It basically fixes the Syskit transformer support, that has been broken circa 2015-2016.

How it happened:
- along with Syskit 2 support, I've added SDF load support to the SDF (ca. 2014-2015)
- this support was rejected #4 
- didn't have the time to refactor the complete gazebo/sdf support at the time
- lost track of this particular issue, so it never happened.

This is the non-SDF bits. To be honest, even though extracting the whole syskit-sdf support in a separate package is on my todo-list, it's pretty low. So, I am expecting this to happen again.

Obviously, no-one is using Syskit or SDF without going through [the Gazebo package set](https://www.rock-robotics.org/rock-and-syskit/), which overrides this repo with the `rock-gazebo` fork.

doudou committed:
syskit: update for the cleanup on Engine in Syskit 2

doudou committed:
syskit: workaround an issue with the transformer used as modelling tool

If a component gets some transformer-related modelling but does not use
the transformer itself at the C++ level, some forms of hell ensues.

There's no way around it short of (finally, and needed) refactoring of
the whole transformer thingy. Work around the problem for now.

doudou committed:
syskit: allow specifying frame assignments for a device's underlying driver

doudou committed:
syskit: hook into Profile::RobotDefinition#inject_di_context to merge the profile's transformer setup

This ensures that requirements generated from device definitions
have a proper transformer setup, which is really necessary now
that the transformer plugin validates the frame names on devices.

doudou committed:
syskit: don't sanitize object IDs

This is really the job of the Syskit graphviz API

doudou committed:
syskit: improve information provided when a chain cannot be found

InvalidChain will now display the transformer configuration

doudou committed:
syskit: adapt to the addition of #associate_ports_to_transform

This declaration is a weaker version of #transform_input.
transform_input declares that a port inputs a certain frame *and*
that this transform is fed manually to the transformer. The new
predicate only declares the association. All transform_input/
transform_output ports have their association declared automatically.
.
each_transform_port now enumerates all associations (instead of
only the transform_input/transform_output ports), as it was its
actual intended use in most cases. Replace the few places where
the intended use was to enumerate the transform input/output by
the correct enumerator (interestingly, they were doing class-based
tests for inputs or outputs, so the change ends up being a cleanup
as well)

doudou committed:
syskit: no need to check if we should validate, we're in validate_abstract_network

doudou committed:
syskit: update validation methods to the latest SystemNetworkGenerator API

doudou committed:
ruby: allow sharing transformation definitions

This commit makes the transform objects immutable so that the
transformation definitions can be shared freely

It improves loading times in Syskit by some 25% just by avoiding
a few tens of thousand of map-copy operations that are there
for nothing.

doudou committed:
syskit: fix Syskit's no-period-for-broadcaster warnings

doudou committed:
syskit: do not use the orogen plugin's transformer method to give access to the transformer setup

This method is really geared towards code generation, which very
often causes headaches. Instead, get access to the extension and
instance_eval ourselves.

doudou committed:
syskit: move the definition of #static_transforms to allow for placeholder tasks

doudou committed:
syskit: use Typelib's new-style Types accessor

doudou committed:
syskit: forward the list of newly-defined definitions in #use_profile

doudou committed:
syskit: use new-style Types access API

doudou committed:
syskit: add and forward use_profile's transform_names argument

doudou committed:
syskit: display available frames when raising InvalidConfiguration

doudou committed:
syskit: fix usage of app lifecycle hooks

doudou committed:
test: refresh the common test setup file

doudou committed:
syskit: fix dangling transformer producer nodes in graphviz generation in some cases

doudou committed:
ruby: add Configuration#rename_frames to rename a frame in a configuration

doudou committed:
syskit: do not export the transformer setup in tests

It's done so often that it starts to actually have a significant cost.

doudou committed:
syskit: do not add an ordering edge between inputs (resp. outputs) if there are none

This should solve the spurious "inputs_${ID}" / "outputs_${ID}" randomly
appearing in a network graph.

doudou committed:
syskit: fix the order between the transformer frames and the ports

The edges where always defined as "from frame to port". However,
even with dir=none, dot will make the edge start at the left side
of an output port to go to the right side of the frame. Fix this.

Moreover, remove the "constraint=false" parameter, which improves
the final layout quite a bit.

doudou committed:
ruby: add #resolve_static_chain to compute a transform between two frames

doudou committed:
syskit: fix writing the properties on the task contexts

It was still directly accessing orocos_task instead of using
the syskit property API. It would (magically ?) work the first time,
but then randomly fail on reconfiguration

doudou committed:
syskit: import the transformer info first in use_profile

The info can be relevant when promoting requirements from the used
profile, so we must make sure it has been integrated first.

doudou committed:
syskit: validate frame selections against the known frames

So far, the frames were checked only when trying to resolve the chains.
This is better. Ideally, we would need to do this validation at loading
time, but that would require a new model-validation hook (or maybe a
model-validation step in tests ?)

doudou committed:
fix Configuration#rename_frames

It was not mapping the names in the transforms and example_transforms
hashes.

doudou committed:
ruby: rename transformation_for and has_transformation? into transform_for and has_transform?

All the other methods use 'transform'

doudou committed:
ruby: remove invalid test

We freeze the transforms, but not the eigen objects (which would require
doing a lot of work within the Eigen bindings)

doudou committed:
syskit: define InstanceRequirements#find_transform_of_port

doudou committed:
ruby: add a Rakefile to ease running the non-Syskit tests

doudou committed:
syskit: fix unit test suite

g-arjones commented on ruby/lib/transformer/syskit/extensions/system_network_generator.rb:
```diff
@@ -0,0 +1,71 @@
+
+            if Syskit.conf.transformer_enabled?
+                plan.find_local_tasks(Syskit::TaskContext).each do |task|
+                    next if !(tr = task.model.transformer)
```
Did we remove the "prefer unless over if not" cop or is this here just because you copied this code from pre-rubocop days?

doudou replied:
Second one. Like beginning of flatfish days.

This is one package that needs a rubocop cleanup

doudou committed:
syskit: do not attempt to load the transformer task library in SyskitPlugin.enable

This is at a place where the common_models bundle is not necessarily
available. The plugin already loads the plugin in require_models (the
right place).

This was added to get the test suite to pass when files were
executed one by one. Just add using_task_library there.

2maz reviewed (changes requested):
Could you add some documentation for the added methods on lib/transformer.rb

doudou committed:
ruby: add documentation

doudou commented:
> Could you add some documentation for the added methods on lib/transformer.rb

Sure. Done.

2maz commented:
> > Could you add some documentation for the added methods on lib/transformer.rb
> 
> Sure. Done.

I think you missed to push the changes

doudou commented:
> I think you missed to push the changes

I mistakenly pushed them to the rock.gazebo repo ...
